## ひろしまQuest2020#stayhome【コース予測部門】

In [218]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import lightgbm as lgb
import gc
%matplotlib inline
pd.set_option('display.max_Columns', 300)

### 管理番号

In [219]:
model_No = '22A'

### sub-modelを使用するかどうか

In [220]:
use_sub_model = True

### n/4

In [221]:
sample_No = 4

In [222]:
ALL_MERGE = 'intermediate/{}/all_merge_{}_{}.f'.format(model_No, model_No, sample_No)
SUBMIT = 'submit/{}/course_{}.csv'.format(model_No, model_No)
SUBMIT_F = 'submit/{}/course_{}_{}.f'.format(model_No, model_No, sample_No)

if use_sub_model:
    ALL_MERGE = 'intermediate/{}/all_merge_{}_{}_sub.f'.format(model_No, model_No, sample_No)
    SUBMIT = 'submit/{}/course_{}_sub.csv'.format(model_No, model_No)
else:
    OUT_SUBMODEL = 'intermediate/{}/course_predict_{}_{}.f'.format(model_No, model_No, sample_No)

In [223]:
all_pitch = pd.read_feather(ALL_MERGE)
all_pitch.shape

(778767, 211)

In [224]:
if use_sub_model:
    all_pitch['predict_curve'] = all_pitch['predict_curve'] / all_pitch['predict_straight']
    all_pitch['predict_slider'] = all_pitch['predict_slider'] / all_pitch['predict_straight']
    all_pitch['predict_shoot'] = all_pitch['predict_shoot'] / all_pitch['predict_straight']
    all_pitch['predict_fork'] = all_pitch['predict_fork'] / all_pitch['predict_straight']
    all_pitch['predict_changeup'] = all_pitch['predict_changeup'] / all_pitch['predict_straight']
    all_pitch['predict_sinker'] = all_pitch['predict_sinker'] / all_pitch['predict_straight']
    all_pitch['predict_cutball'] = all_pitch['predict_cutball'] / all_pitch['predict_straight']
    all_pitch.drop(columns=[
        'predict_0', 'predict_1', 'predict_2', 'predict_3', 'predict_4', 'predict_5', 'predict_6',
        'predict_7', 'predict_8', 'predict_9', 'predict_10', 'predict_11', 'predict_12',
       'predict_straight'
    ], inplace=True)

In [225]:
print(all_pitch.shape)
column_cnt = len(all_pitch.columns)
all_pitch.head()

(778767, 197)


No  ball  course  pitch_cnt_in_game  inning  bat_cnt_in_inning  \
0   0   0.0     0.0                  1       1                  1   
1   1   0.0     8.0                  2       1                  1   
2   2   0.0     5.0                  3       1                  1   
3   3   0.0    12.0                  4       1                  2   
4   4   0.0     8.0                  5       1                  2   

   pitch_cnt_in_bat  pitch_order  player_cnt_in_game  pitcher_cnt_in_game  \
0                 1            1                   1                    1   
1                 2            1                   1                    2   
2                 3            1                   1                    3   
3                 1            1                   2                    4   
4                 2            1                   2                    5   

   pitcher_cnt_in_inning  bat_order  bat_cnt_in_game  home_point  away_point  \
0                      1          1                1           0           0   
1                      2          1                1           0           0   
2                      3          1                1           0           0   
3                      4          2                1           0           0   
4                      5          2                1           0           0   

   out_cnt  ball_cnt  strike_cnt  pitch_LR  bat_LR  bc_curve  bc_slider  \
0        0         0           0       0.0     1.0  0.237468   0.325101   
1        0         0           1       0.0     1.0  0.210923   0.247198   
2        0         0           2       0.0     1.0  0.138758   0.130040   
3        1         0           0       0.0     0.0  0.213866   0.474931   
4        1         0           1       0.0     0.0  0.215241   0.420099   

   bc_shoot   bc_fork  bc_changeup  bc_sinker  bc_cutball  bc_high_str  \
0  0.209547  0.145227     0.092978   0.027829    0.113620     0.135883   
1  0.177502  0.320359     0.147952   0.039943    0.116772     0.106283   
2  0.092263  0.419906     0.117327   0.038140    0.046858     0.076007   
3  0.139964  0.088519     0.038543   0.007023    0.138168     0.151308   
4  0.145723  0.196762     0.051566   0.009680    0.118444     0.113468   

   bc_high_ball  bc_mid_str  bc_low_str  bc_low_ball  bc_left_str  \
0      0.221489    0.204959    0.174254     0.263415     0.242045   
1      0.230235    0.157397    0.138195     0.367890     0.184621   
2      0.284249    0.102015    0.095055     0.442674     0.118315   
3      0.196937    0.206965    0.157643     0.287147     0.105057   
4      0.206297    0.162901    0.127988     0.389347     0.084346   

   bc_left_ball  bc_center_str  bc_right_str  bc_right_ball  first  second  \
0      0.294634       0.157681      0.115370       0.190270      0       0   
1      0.355720       0.111331      0.105923       0.242405      0       0   
2      0.488095       0.076190      0.078571       0.238828      0       0   
3      0.153718       0.164756      0.246104       0.330367      0       0   
4      0.162167       0.120564      0.199445       0.433477      0       0   

   third  base_cnt  top_bot  role  pos_pit  date_from_opening  elapsed_min  \
0      0         0        0     1        0                  0     0.000000   
1      0         0        0     1        0                  0     0.200000   
2      0         0        0     1        0                  0     0.383333   
3      0         0        0     1        0                  0     1.016667   
4      0         0        0     1        0                  0     1.200000   

   min_diff  total_inning  elapsed_from_inning  ballnum_from_inning  \
0       NaN             0             0.000000                    1   
1  0.200000             0             0.200000                    2   
2  0.183333             0             0.383333                    3   
3  0.633333             0             1.016667                    4   
4  0.183333             0             1.

### train
- 行数: 257117

In [226]:
train = all_pitch.dropna(subset=['course'])
train.shape

(257117, 197)

#### 特徴量に加えた部分を抜く
- train_pitch_1 = train_pitch.query('index <= 60000')
- train_pitch_2 = train_pitch.query('index > 60000 & index <= 120000')
- train_pitch_3 = train_pitch.query('index > 120000 & index <= 180000')
- train_pitch_4 = train_pitch.query('index > 180000')

In [227]:
if sample_No == 1:
    train = train.query('index > 60000')
elif sample_No == 2:
    train = train.query('index <= 60000 | index > 120000')
elif sample_No == 3:
    train = train.query('index <= 120000 | index > 180000')
elif sample_No == 4:
    train = train.query('index <= 180000')
train.shape

(180001, 197)

### test
- 行数: 521650

In [228]:
test = all_pitch[all_pitch['course'].isnull()]
test.shape

(521650, 197)

In [229]:
del all_pitch
gc.collect()

47

In [230]:
train_d = train.drop([
    'No', 
    'course', 
    'ball'
], axis=1)

In [231]:
test_d = test.drop([
    'No', 
    'course', 
    'ball'
], axis=1)

### LGBM学習
- multiclass : クラス数=13

In [232]:
lgb_param = {
    'objective' : 'multiclass',
    'boosting_type': 'gbdt',
    'metric' : 'multi_logloss',
    'num_class' : 13,
    'seed' : 0,
    'learning_rate' : 0.1,
    'lambda_l1': 8.769293390201968, 
    'lambda_l2': 3.913949617576324e-05, 
    'num_leaves': 6, 
    'feature_fraction': 0.4, 
    'bagging_fraction': 0.8391111798378441, 
    'bagging_freq': 4, 
    'min_child_samples': 50
}

In [233]:
def lightgbm_cv(lgb_param, lgb_train):
    cv_results = lgb.cv(lgb_param, lgb_train,
                    num_boost_round=15000,
                    early_stopping_rounds=100,
                    verbose_eval=100,
                    nfold=4)

    num_boost_round = len(cv_results['multi_logloss-mean'])
    print('Best num_boost_round:', num_boost_round)
    best_cv_score = cv_results['multi_logloss-mean'][-1]
    print('Best CV score:', best_cv_score)
    best_iter = int(num_boost_round * 1.1)
    return best_iter

### Feature Importance

In [234]:
def feature_importance(lgb_model):
    fi = lgb_model.feature_importance()
    fn = lgb_model.feature_name()
    df_feature_importance = pd.DataFrame({'feat_name':fn, 'feat_imp':fi})
    df_feature_importance.sort_values('feat_imp', inplace=True)
    return df_feature_importance

In [235]:
%%time
lgb_train = lgb.Dataset(train_d, train['course'])
best_iter = lightgbm_cv(lgb_param, lgb_train)

[100]	cv_agg's multi_logloss: 2.34142 + 0.00166748
[200]	cv_agg's multi_logloss: 2.33402 + 0.00177067
[300]	cv_agg's multi_logloss: 2.33095 + 0.00188401
[400]	cv_agg's multi_logloss: 2.32941 + 0.00197108
[500]	cv_agg's multi_logloss: 2.32854 + 0.00207544
[600]	cv_agg's multi_logloss: 2.32805 + 0.00226962
[700]	cv_agg's multi_logloss: 2.32788 + 0.0022847
[800]	cv_agg's multi_logloss: 2.32783 + 0.00240046
Best num_boost_round: 772
Best CV score: 2.3278100071814736
CPU times: user 1h 3min 5s, sys: 13min 3s, total: 1h 16min 8s
Wall time: 4min 46s


In [236]:
%%time
lgb_model = lgb.train(lgb_param, lgb_train, num_boost_round=best_iter)

CPU times: user 18min 57s, sys: 3min 47s, total: 22min 44s
Wall time: 1min 25s


In [237]:
feature_importance(lgb_model).tail(30)

feat_name  feat_imp
178  salary_year_dif_b-c       509
103               slider       511
118             left_str       513
6    pitcher_cnt_in_game       535
119            left_ball       540
140           batter_cnt       553
113             high_str       554
116              low_str       556
121            right_str       563
166          bmi_dif_p-b       579
122           right_ball       585
45   elapsed_from_inning       590
114            high_ball       601
182          bmi_dif_b-c       610
120           center_str       613
50    ave_elapsed_batter       625
51   diff_elapsed_batter       709
43              min_diff       731
52   diff_elapsed_inning       804
53     diff_elapsed_game       829
49    ave_elapsed_inning       883
192       predict_sinker       917
48      ave_elapsed_game       938
187        predict_curve       985
188       predict_slider      1023
41     date_from_opening      1029
193      predict_cutball      1029
191     predict_changeup      1092
189        predict_shoot      1356
190         predict_fork      1361

### 予測

In [238]:
%%time
predict = lgb_model.predict(test_d, num_iteration = lgb_model.best_iteration)

CPU times: user 5min 56s, sys: 2.39 s, total: 5min 58s
Wall time: 24.1 s


## 結果出力

In [239]:
submit = pd.DataFrame(predict)
submit.reset_index(inplace=True)
print(submit.shape)
submit.head()

(521650, 14)


index         0         1         2         3         4         5  \
0      0  0.063018  0.141902  0.074304  0.038316  0.066545  0.048542   
1      1  0.046781  0.118927  0.070143  0.028923  0.055281  0.040132   
2      2  0.029837  0.085635  0.061242  0.020259  0.040105  0.028776   
3      3  0.060481  0.107783  0.058548  0.026200  0.037187  0.039360   
4      4  0.045417  0.096239  0.070652  0.031646  0.044110  0.052813   

          6         7         8         9        10        11        12  
0  0.045556  0.053290  0.037147  0.153603  0.086022  0.089191  0.102564  
1  0.042239  0.086690  0.041301  0.124430  0.105603  0.108243  0.131306  
2  0.052938  0.070792  0.029188  0.120732  0.126320  0.172043  0.162136  
3  0.039726  0.052520  0.035873  0.180331  0.108795  0.112649  0.140546  
4  0.055773  0.102178  0.047582  0.099027  0.094035  0.109502  0.151027

### CSV出力

In [240]:
#submit.to_csv(SUBMIT, header=False, index=False)

### Feather出力

In [241]:
submit_f = submit.drop(columns=['index'])
submit_f.rename(columns={
        0: 'predict_0', 1: 'predict_1', 2: 'predict_2', 3: 'predict_3',
        4: 'predict_4', 5: 'predict_5', 6: 'predict_6', 7: 'predict_7',
        8: 'predict_8', 9: 'predict_9', 10: 'predict_10', 11: 'predict_11', 12: 'predict_12'
    }, inplace=True)
submit_f.to_feather(SUBMIT_F)
print(SUBMIT_F)

submit/22A/course_22A_4.f


### 結果まとめ

In [242]:
ave_result = (sample_No == 4)

In [243]:
if ave_result:
    result = 'submit/{}/course_{}_{}.f'.format(model_No, model_No, 1)
    print(result)
    df = pd.read_feather(result)
    columns = ['predict_0', 'predict_1', 'predict_2', 'predict_3', 
               'predict_4', 'predict_5', 'predict_6', 'predict_7',
              'predict_8', 'predict_9', 'predict_10', 'predict_11', 'predict_12']

    for i in range(2,5):
        result = 'submit/{}/course_{}_{}.f'.format(model_No, model_No, i)
        print(result)
        temp = pd.read_feather(result)
        for c in columns:
            df[c] = df[c] + temp[c]

    for c in columns:
        df[c] = df[c]/4

    # 出力
    df = df.reset_index()
    df.to_csv(SUBMIT, header=False, index=False)
    print(SUBMIT)

submit/22A/course_22A_1.f
submit/22A/course_22A_2.f
submit/22A/course_22A_3.f
submit/22A/course_22A_4.f
submit/22A/course_22A_sub.csv


### 球種予測で使用する

In [244]:
%%time
if not use_sub_model:
    train_predict = lgb_model.predict(train_d, num_iteration = lgb_model.best_iteration)

CPU times: user 4 µs, sys: 0 ns, total: 4 µs
Wall time: 7.87 µs


In [245]:
if not use_sub_model:
    df_train_predict = pd.DataFrame(train_predict).reset_index()
    submodel = pd.concat([df_train_predict, submit], ignore_index=True)
    submodel.drop(columns=['index'], inplace=True)
    submodel.rename(columns={
        0: 'predict_0', 1: 'predict_1', 2: 'predict_2', 3: 'predict_3',
        4: 'predict_4', 5: 'predict_5', 6: 'predict_6', 7: 'predict_7',
        8: 'predict_8', 9: 'predict_9', 10: 'predict_10', 11: 'predict_11', 12: 'predict_12'
    }, inplace=True)
    print(submodel.shape)
    submodel.head()

In [246]:
if not use_sub_model:
    submodel.to_feather(OUT_SUBMODEL)
    print(OUT_SUBMODEL)

### SIGNATE

In [247]:
'signate submit --competition-id=276 ./{} --note feat={}'.format(SUBMIT, column_cnt)

'signate submit --competition-id=276 ./submit/22A/course_22A_sub.csv --note feat=197'

### GCS

In [248]:
'gsutil cp ./{} gs://yasunobu_baseball/submit/'.format(SUBMIT)

'gsutil cp ./submit/22A/course_22A_sub.csv gs://yasunobu_baseball/submit/'

### 評価結果
#### 22A_sub(197) ... 22相当に戻す
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- (1) Best CV score: 2.341708495128639
- (2) Best CV score: 2.338236763777662
- (3) Best CV score: 2.336293070462755
- (4) Best CV score: 2.3278100071814736
- 評価結果  : 2.3511614

#### 22A(190) ... 22相当に戻す
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- (1) Best CV score: 2.3407026780391287
- (2) Best CV score: 2.3386571769120112
- (3) Best CV score: 2.3406188200447495
- (4) Best CV score: 2.3414496657868056
- 評価結果  : 2.3485811 ★

#### 23_tune(190) 
- 'learning_rate' : 0.1,
- 'num_leaves': 4,
- (1) Best CV score: 2.341850353687299
- (2) Best CV score: 2.3396210433383273
- (3) Best CV score: 2.341554253259382
- (4) Best CV score: 2.3425899300822706
- 評価結果  : 2.3493330

#### 23(190) ... 2017年の集計がない選手はNaNに
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- (1) Best CV score: 2.3410495769108106
- (2) Best CV score: 2.3389861210773724
- (3) Best CV score: 2.3409606333569144
- (4) Best CV score: 2.3417060924636233
- 評価結果  : 2.3486626

#### 22_sub3(198) ... 球種の予測をそのまま使う
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- (1) Best CV score: 2.324145615655472
- (2) Best CV score: 2.320827785286294
- (3) Best CV score: 2.324029135378343
- (4) Best CV score: 2.3253020291900324
- 評価結果  : 2.3523530

#### 22_sub2(197) ... 特徴量で集計した部分を抜く
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- (1) Best CV score: 2.327086859587935
- (2) Best CV score: 2.3242695883810653
- (3) Best CV score: 2.326650154544137
- (4) Best CV score: 2.3279934162416356
- 評価結果  : 2.3510544

#### 22_sub(197)
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- (1) Best CV score: 2.320717948163731
- (2) Best CV score: 2.3204691031051947
- (3) Best CV score: 2.3217573750471856
- (4) Best CV score: 2.320525945710633
- 評価結果  : 2.3517300

#### 22(190)
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- (1) Best CV score: 2.3350431782409524
- (2) est CV score: 2.335302207655757
- (3) Best CV score: 2.3353277079714583
- (4) Best CV score: 2.3340380616284415
- 評価結果  : 2.3487659

#### 21_sub(198)
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 639
- Best CV score: 2.3159487785697754
- Wall time: 5min
- 評価結果  : 2.3572586

#### 20_sub(173)
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 793
- Best CV score: 2.315349281506766
- Wall time: 9min 57s
- 評価結果  : 2.3585163

#### 20(166) ... player同士の組み合わせを削除
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 572
- Best CV score: 2.3271237073894153
- Wall time: 6min 59s
- 評価結果  : 2.3552697

#### 19_sub(197) 
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 774
- Best CV score: 2.316818574234304
- Wall time: 10min 49s
- 評価結果  : 2.3581029

#### 19(190) 
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 550
- Best CV score: 2.327580481542456
- Wall time: 7min 36s
- 評価結果  : 2.3554862

#### 18_sub(225) ...サブモデルで球種の予測を追加
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 548
- Best CV score: 2.308116414786115
- Wall time: 8min 25s
- 評価結果  : 2.3638683

#### 18(217) 
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 373
- Best CV score: 2.3204478805445055
- Wall time: 6min 9s
- 評価結果  : 2.3604751

#### 17(192) 
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 415
- Best CV score: 2.3200530570975535
- Wall time: 15min 10s
- 評価結果  : 2.3607863

#### 16(181)  ...選手IDごとの2017年のコース実績
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 355
- Best CV score: 2.3203553166835866
- Wall time: 5min 8s
- 評価結果  : 2.3608998

#### 15(153)  ...左右で球種/コースを別に集計
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 932
- Best CV score: 2.3387434142884445
- Wall time: 9min 41s
- 評価結果  : 2.3552543

#### 14(153)  ...左右で球種を別に集計
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 909
- Best CV score: 2.340082741044867
- Wall time: 9min 40s
- 評価結果  : 2.3555548

#### 3(157)  ...前の投球からの時間差を追加
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 857
- Best CV score: 2.340703597440808
- Wall time: 9min 20s
- 評価結果  : 2.3554087

#### 2(156)  ...merge12のパラメータをチューニング
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 835
- Best CV score: 2.3419206656914
- Wall time: 9min 7s
- 評価結果  : 2.3555299

#### 2(156)  ...merge12(2017のコースデータを付加) 
- 'learning_rate' : 0.1,
- Best num_boost_round: 121
- Best CV score: 2.344104226715559
- Wall time: 12min 30s
- 評価結果  : 2.3564326

#### 1(143)  ...球種予測の11と同じデータ 
- 'learning_rate' : 0.1,
- Best num_boost_round: 122
- Best CV score: 2.3445089434874253
- Wall time: 12min 2s (local)
- 評価結果  : 2.3562648